### PR TITLE
Use Struct Field Resolver instead of Method

### LINKED ISSUE #28: Use field as resolver instead of method.
Would it be possible to use a struct field as a resolver instead of always using methods? Most of my methods are just simple getters so it would be convenient to allow resolver to default to field. Then I could implement method only when I need special logic.

### PR DESCRIPTION
**Pull request by salman-ahmad**
This fixes #28. Essentially, it reduces lot of boilerplate code by making use of struct field as resolvers instead of creating methods for each struct field. 

Now, methods are required **only when**

1. a struct implements a GraphQL interface. In this case, you create methods for fields which overlap
1. a struct exposes a GraphQL field that accepts additional arguments i.e., `first: Int, last: Int` etc. 
1. there is GraphQL `union` type

By using struct fields as resolvers, one could also benefit from DRY codebase by not having a pair of a struct which look the same (one for GraphQL and one for DB)

### Does this break existing API? 

No. This change is completely transparent and you can continue using your existing codebase. 

### Can I still continue using methods as resolvers?

Yes, by default, it uses method resolvers. Even when `UseFieldResolvers()` is turned on, the method is given higher priority than the field. When there is no method, the field will be used if there is any. 

### How can I use struct fields as resolvers?

When invoking `graphql.ParseSchema()` or `graphql.MustParseSchema()` to parse the schema, you pass in `graphql.UseFieldResolvers()` as an argument. For an example, take a look at `./example/social/server/server.go`

### Additional Notes

- for an example, take a look at `./example/social/social.go` & `./example/social/server/server.go`
- Unit tests are missing

### CODE REVIEW

**Comment by bazaglia** [comment]
@pavelnikolov are you OK with merging this?

**Comment by nanozuki** [comment]
I think this is a really useful feature and can't wait to use it. What't the plan for merging?

**Comment by pavelnikolov** [comment]
I'll review the PR shortly. I'm sorry for the delay on this.

**Comment by eveld** [comment]
Can't wait to throw away half of my resolver code after this is merged! :)

**Comment by jenting** [comment]
Any update for this PR be merged?

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,6 @@
 /internal/validation/testdata/graphql-js
 /internal/validation/testdata/node_modules
 /vendor
+.DS_Store
+.idea/
+.vscode/
```
Comment (pavelnikolov) [inline .gitignore]: Unrelated change! Please, restore this file - these need to be in your global ignore file.

Comment (salman-ahmad) [reply .gitignore]: I have restored this file

**Comment by salman-ahmad** [comment]
@pavelnikolov I have addressed your review comment. Please merge this PR, thank you1

**Comment by ItsLeeOwen** [comment]
Thank the gods, and @salman-ahmad  I was just looking for this functionality.  ❤️

**Comment by pavelnikolov** [comment]
> I would suggest that graphql.UseFieldResolvers() possibly become the default.

@ItsLeeOwen , your suggestion would introduce a breaking change which I would really like to avoid. Instead, this feature will be opt-in.

**Comment by ItsLeeOwen** [comment]
@pavelnikolov   Cool, I'm curious what would that breaking change be?

**Comment by pavelnikolov** [comment]
@ItsLeeOwen oh, my bad 🤦‍♂️ 
There is no breaking change. I got confused..

File: README.md [hunk 2]
```diff
@@ -65,7 +65,17 @@ $ curl -XPOST -d '{"query": "{ hello }"}' localhost:8080/query
...
 
-A resolver must have one method for each field of the GraphQL type it resolves. The method name has to be [exported](https://golang.org/ref/spec#Exported_identifiers) and match the field's name in a non-case-sensitive way.
+A resolver must have one method or field for each field of the GraphQL type it resolves. The method or field name has to be [exported](https://golang.org/ref/spec#Exported_identifiers) and match the schema's field's name in a non-case-sensitive way.
+You can use struct fields as resolvers by using `SchemaOpt: UseFieldResolvers()`. For example,
+```
+opts := []graphql.SchemaOpt{graphql.UseFieldResolvers()}
+schema := graphql.MustParseSchema(s, &query{}, opts...)
+```   
+
+When using `UseFieldResolvers`, a field will be used *only* when:
```
Comment (pavelnikolov) [inline README.md]: The above needs to be clarified a little, for example:
> When using `UseFieldResolvers` schema option, a struct field will be used *only* when:

File: README.md [hunk 3]
```diff
@@ -65,7 +65,17 @@ $ curl -XPOST -d '{"query": "{ hello }"}' localhost:8080/query
...
+- there is no method
+- it does not implement an interface
```
Comment (pavelnikolov) [inline README.md]: > it does not implement an interface

What does "it" stand for in this sentence? I understand it but a first-time user of this library might not understand this. Can you, please, put a more descriptive sentence?

File: README.md [hunk 4]
```diff
@@ -65,7 +65,17 @@ $ curl -XPOST -d '{"query": "{ hello }"}' localhost:8080/query
...
+- it does not have arguments
```
Comment (pavelnikolov) [inline README.md]: Same here - "it" is ambiguous for new starters.

File: graphql.go [hunk 5]
```diff
@@ -84,6 +83,13 @@ func UseStringDescriptions() SchemaOpt {
 	}
 }
 
+// Specifies whether to use struct field resolvers
```
Comment (pavelnikolov) [inline graphql.go]: Comments need to start with the func/method name
https://github.com/golang/go/wiki/CodeReviewComments#comment-sentences

Please, use:
```
UseFieldResolvers...
```

**Comment by pavelnikolov** [comment]
Thank you for your contribution!

**Comment by tobstarr** [comment]
:tada: just at the right time! Thank you so much!

**Comment by marcelvanwaaijen** [comment]
Question about the social example: What is the use of the FriendsResolver method? When is it called? I put some debug logging in it, but that never shows, also the pagination doesn't seem to be doing anything currently?

**Comment by pavelnikolov** [comment]
@marcelvanwaaijen I haven't looked into it in details but my wild guess is that it was an existing example and when this feature was implemented we ended up adding `graphql.UseFieldResolvers()` which uses fields instead of methods. Then we started using the fields and stopped using the method. I would suggest opening a separate issue if this is the case as this one was related to adding the field resolvers feature, which is already done.
